cli/util/synchronization.rs

```diff
@@ -0,0 +1,27 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+use tokio::sync::Semaphore;
+use tokio::sync::SemaphorePermit;
+
+pub struct SingleConcurrencyEnforcerPermit<'a>(SemaphorePermit<'a>);
+
+/// Enforces only one request can enter the section of code
+/// that holds a permit.
+#[derive(Debug)]
+pub struct SingleConcurrencyEnforcer(Semaphore);
+
+impl Default for SingleConcurrencyEnforcer {
+  fn default() -> Self {
+    Self(Semaphore::new(1))
+  }
+}
+
+impl SingleConcurrencyEnforcer {
+  /// Acquire a permit to force other calls to `acquire` to wait until the
+  /// permit is dropped. Permits are acquired in order (first-in, first-out).
```

Comment:
I just recalled looking at TaskQueue, that this is actually not true with semaphore from my previous testing (it's not FIFO... it's only mostly FIFO)